sdks/java/core/src/main/java/org/apache/beam/sdk/metrics/MetricResults.java

```diff
@@ -31,4 +33,69 @@
    * Query for all metrics that match the filter.
    */
   public abstract MetricQueryResults queryMetrics(MetricsFilter filter);
+
+  // Matching logic is implemented here rather than in MetricsFilter because we would like
+  // MetricsFilter to act as a "dumb" value-object, with the possibility of replacing it with
+  // a Proto/JSON/etc. schema object.
+  protected boolean matches(MetricsFilter filter, MetricKey key) {
+    return filter == null
+        || (matchesName(key.metricName(), filter.names())
+        && matchesScope(key.stepName(), filter.steps()));
+  }
+
+  /**
+   * {@code subPathMatches(haystack, needle)} returns true if {@code needle}
+   * represents a path within {@code haystack}. For example, "foo/bar" is in "a/foo/bar/b",
+   * but not "a/fool/bar/b" or "a/foo/bart/b".
+   */
+  public boolean subPathMatches(String haystack, String needle) {
```

Comment:
It would be better to make these static helper methods or available on a separate class. The reason being that `MetricResults` is a user-facing part of the API, so we shouldn't have public methods in it that aren't meant to be user consumed (such as `subPathMatches`)